flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSourceReader.java

```diff
@@ -0,0 +1,213 @@
+        for (SourceSplit split : state) {
+            wrappedSplits.add(new HybridSourceSplit<>(readerIndex, split));
+        }
+        return wrappedSplits;
+    }
+
+    public static <SplitT extends SourceSplit> List<SplitT> unwrappedSplits(
+            List<HybridSourceSplit<SplitT>> splits) {
+        List<SplitT> unwrappedSplits = new ArrayList<>(splits.size());
+        for (HybridSourceSplit<SplitT> split : splits) {
+            unwrappedSplits.add(split.getWrappedSplit());
+        }
+        return unwrappedSplits;
+    }
+
+    @Override
+    public CompletableFuture<Void> isAvailable() {
+        return availabilityFuture = currentReader.isAvailable();
+    }
+
+    @Override
+    public void addSplits(List<HybridSourceSplit<?>> splits) {
+        LOG.info(
+                "Adding splits subtask={} sourceIndex={} {} {}",
+                readerContext.getIndexOfSubtask(),
+                currentSourceIndex,
+                currentReader,
+                splits);
+        List<SourceSplit> realSplits = new ArrayList<>(splits.size());
+        for (HybridSourceSplit<?> split : splits) {
+            Preconditions.checkState(
+                    split.sourceIndex() == currentSourceIndex,
+                    "Split %s while current source is %s",
+                    split,
+                    currentSourceIndex);
+            realSplits.add(split.getWrappedSplit());
+        }
+        currentReader.addSplits((List) realSplits);
+    }
+
+    @Override
+    public void notifyNoMoreSplits() {
+        currentReader.notifyNoMoreSplits();
+        LOG.debug(
+                "No more splits for reader subtask={} sourceIndex={} {}",
+                readerContext.getIndexOfSubtask(),
+                currentSourceIndex,
+                currentReader);
+    }
+
+    @Override
+    public void handleSourceEvents(SourceEvent sourceEvent) {
+        if (sourceEvent instanceof SwitchSourceEvent) {
+            SwitchSourceEvent sse = (SwitchSourceEvent) sourceEvent;
+            LOG.debug(
+                    "Switch source event: subtask={} sourceIndex={}",
+                    readerContext.getIndexOfSubtask(),
+                    sse.sourceIndex());
+            setCurrentReader(sse.sourceIndex());
+            if (availabilityFuture != null && !availabilityFuture.isDone()) {
+                // continue polling
+                availabilityFuture.complete(null);
+            }
+        } else {
+            currentReader.handleSourceEvents(sourceEvent);
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        currentReader.close();
+        LOG.debug(
+                "Reader closed: subtask={} sourceIndex={} {}",
+                readerContext.getIndexOfSubtask(),
+                currentSourceIndex,
+                currentReader);
+    }
+
+    private void setCurrentReader(int index) {
+        Preconditions.checkState(
```

Comment:
should the assertion be strictly `currentSourceIndex == index || currentSourceIndex - 1 == index`?